QuickOpen API enhancements

azatsarynnyy opened the pull request:
closes #1680

This PR extends the Quick Open API to allow the Quick Open widget handle special prefixes which trigger one of the registered handlers. The handlers can be contributed by the extensions. In this PR I've added several handlers:
- `>` - triggers command quick open;
- `#` - triggers workspace symbol quick open;
- `task ` - triggers task quick open;
- `...` - default handler (used without prefix) triggers open file.

Short demo:
![quickopen](https://user-images.githubusercontent.com/1636395/44285889-ae0da700-a26f-11e8-9439-12298ee40203.gif)

akosyakov commented on packages/core/src/browser/quick-open/quick-open-service.ts:
```diff
@@ -57,12 +63,257 @@ export namespace QuickOpenOptions {
+     * the quick open widget matches this handler's prefix.
+     * Allows to initialize the model with some initial data.
+     */
+    init: () => void;
```
`init(): void;` and for other methods?

akosyakov commented on packages/core/src/browser/quick-open/quick-open-service.ts:
```diff
@@ -57,12 +63,257 @@ export namespace QuickOpenOptions {
+
+    protected readonly handlers: Map<string, QuickOpenHandler> = new Map();
+    protected readonly toDispose = new DisposableCollection();
+    protected defaultHandler: QuickOpenHandler;
```
`QuickOpenHandler | undefined` the same for `getDefaultHandler`

akosyakov commented on packages/core/src/browser/quick-open/quick-open-service.ts:
```diff
@@ -57,12 +63,257 @@ export namespace QuickOpenOptions {
     /**
+     * The internal model which is used when the quick open widget is opened in handling prefix mode.
+     */
+    protected model: QuickOpenModel;
```
How about we built the prefix-based quick open service as a separate class on the top of `QuickOpenService`? This code is not needed for other quick-open services. I would move new APIs to a separate file to leave low-level API uncluttered.

azatsarynnyy commented:
@akosyakov thanks for the review. I've fixed your notes and moved the prefix-based quick open service on the top of the `QuickOpenService`. Nice idea!

akosyakov commented:
@azatsarynnyy great work!

Unfortunately, `workspace symbols` and `go to symbol` with `#` and `@` does not work for me with these changes. Using the dedicated palettes seems to work.

![symbols](https://user-images.githubusercontent.com/3082655/44505664-ed3e5c80-a6a2-11e8-8180-f8b57a9dc4d4.gif)

azatsarynnyy commented:
@akosyakov sounds very strange. I've just checked it several times and everything works perfectly for me. Maybe it's because searching a workspace symbol requires some time. On my machine, for Theia sources, it requires a couple of seconds. Could you, please, check `go to workspace symbol` command again taking this into account?

Note, there's no handler for `@` in this PR. I've no idea yet how to implement it since `go to symbol` is Monaco's internal command and we don't have a quick open model for it. I need to investigate it more.

akosyakov commented:
@azatsarynnyy sorry for the noise with `go to workspace symbol`, forgot that I have to type something :)

please open a follow-up issue about `go to symbol`

akosyakov commented on packages/file-search/src/browser/quick-file-open.ts:
```diff
@@ -69,17 +69,38 @@ export class QuickFileOpenService implements QuickOpenModel {
         const keybinding = this.getKeyCommand();
         if (keybinding) {
-            placeholderText += ` (Press ${keybinding} to show/hide ignored files)`;
+            description += ` (Use ${keybinding} to show/hide ignored files)`;
```
@azatsarynnyy in the case if one is using `Open File...` directly, not via the command palette, it should work as before

azatsarynnyy commented:
@akosyakov @svenefftinge the message `Type '?' to get help on the actions you can take from here` will be shown in case no placeholder has been provided via the options. Otherwise, the provided placeholder is shown. So, `Open File...` works as before.
@akosyakov here's the issue for `Go to Symbol...` prefix - https://github.com/theia-ide/theia/issues/2670
Guys thank you for the PR review!
